### ISSUE TITLE
Mirroring ghcr.io image returns 400 MANIFEST_INVALID (operation unsupported)

### ISSUE THREAD

**Issue by alexbiehl** [distribution/distribution#3530]
When pulling an image hosted on ghcr.io through a Registry set up as a pull-through cache I get a 400 error. 

With debug logs turned on, I see

```
  "err.code": "manifest invalid",
  "err.detail": "operation unsupported",
  "err.message": "manifest invalid",
  ... 
```

The REGISTRY_PROXY_REMOTEURL is pointing to `https://ghcr.io`. I have tried various images .e.g. linkerd/proxy, linuxserver/booksonic-air

I couldn't find anything on the issue tracker so far. I hope it's a known issue with a workaround.

**Comment by milosgajdos** [comment]
We would have to see the snippet of the actual manifest before making/giving any sort of opinion.

What would also be helpful is telling us what version of registry [proxy] you are running etc.

**Comment by alexbiehl** [comment]
@milosgajdos Thanks for the quick response!

> What would also be helpful is telling us what version of registry [proxy] you are running etc.

We are running `registry:2.7.1`. We are running a proxy registry for DockerHub too (some config modulo REGISTRY_PROXY_REMOTEURL and login of course) on which we haven't observed the phenomenon so far.

> We would have to see the snippet of the actual manifest before making/giving any sort of opinion.

The following is the manifest for `ghcr.io/linuxserver/booksonic-air:2009.1.0`: 

```json
% docker manifest inspect ghcr.io/linuxserver/booksonic-air:2009.1.0       
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2206,
         "digest": "sha256:7da5d5e4f2420c329aa38477238ffe2b55b8a776779217dc6912eccb6ffce774",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2206,
         "digest": "sha256:10127f893e7547a7a96327c065e6d4bd7d7a8537303b445fd480bee98716e46e",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2206,
         "digest": "sha256:e5c19b697aeaeabc48df259a62a0c79c52c20676fdc4d7b06a26c5e23a919364",
         "platform": {
            "architecture": "arm64",
            "os": "linux",
            "variant": "v8"
         }
      }
   ]
}
```

**Comment by milosgajdos** [comment]
Interesting. The manifest looks ok to me 🤔 , but I suspect the lack of detail in the error is due to how distribution serialises error into JSON before sending them down the pipes to the consumers.

I have a bit of a theory -- sorry haven't had time to verify this -- but here goes...

Distribution calls `ServeJson` (see [here](https://github.com/distribution/distribution/blob/release/2.7/registry/api/errcode/handler.go#L11)) when serialising the errors (see [here](https://github.com/distribution/distribution/blob/release/2.7/registry/handlers/app.go#L731)).

Now, when the  `error` `Detail` is of type error (such as [here](https://github.com/distribution/distribution/blob/release/2.7/registry/handlers/manifests.go#L310) -- this is one of the possible places where the source of the failure you are seeing might be) we do not marshal the error properly leaving the API client in limbo. (this is the [error marshal code](https://github.com/distribution/distribution/blob/release/2.7/registry/api/errcode/errors.go#L230))

This is a bit unfortunate due to `Go` not being able to marshal error at all. See [here](https://github.com/golang/go/issues/10748).

I'm fairly sure if we do something like this in the `MarshalJSON` for `Errors` type code, that will set the `Detail` of the error to the error _content_ (i.e. the `string`) rather than the `error` (_type_) that gets "swallowed" by `Go`:
```Go
if detail, ok := tmpErr.Detail.(error); ok {
	tmpErr.Detail = detail.Error()
}
```

That will give us more info about the failure. Feel free to open a patch, otherwise I can do it at some point later this week. I had a look at the `main` branch and it seems to be "broken" there too, unfortunately.

**Comment by alexbiehl** [comment]
Ok, I built locally and followed it into manifest/schema1/config_builder.go, specifically:

```go
// emptyTar pushes a compressed empty tar to the blob store if one doesn't
// already exist, and returns its blobsum.
func (mb *configManifestBuilder) emptyTar(ctx context.Context) (digest.Digest, error) {
```

where it fails at 

```go
	// Add gzipped empty tar to the blob store
	descriptor, err = mb.bs.Put(ctx, "", gzippedEmptyTar)
``` 

the returned `err` indeed is `operation unsupported`. 

It turns out that Registries running in proxy mode [do not support pushing images](https://docs.docker.com/registry/configuration/#proxy). Could it be that for some reason the manifest makes us follow a code path that uses the same mechanics as pushing an image to populate it's pull-through cache?

**Comment by alexbiehl** [comment]
And indeed if I replace the occurrences of

```go
		if blobsum, err = mb.emptyTar(ctx); err != nil {
			return nil, err
		}
```

with

```go
    blobsum = digestSHA256GzippedEmptyTar
```

I get my beautiful manifest:

```json
% curl localhost:5000/v2/linuxserver/booksonic-air/manifests/2009.1.0
{
   "schemaVersion": 1,
   "name": "linuxserver/booksonic-air",
   "tag": "2009.1.0",
   "architecture": "amd64",
   "fsLayers": [
   ...
```

**Comment by alexbiehl** [comment]
@milosgajdos I am proposing a fix in https://github.com/distribution/distribution/pull/3532. Feel free to turn down if you see a problem with it.

**Comment by milosgajdos** [comment]
> It turns out that Registries running in proxy mode [do not support pushing images](https://docs.docker.com/registry/configuration/#proxy).

Yes, this has been my default assumption. In the original report you described the issue you were experiencing was when doing PULL; not PUSH 🤔 

Besides, after reading your updates, I'm going to stop here and go on limb to ask you to stop using `v1` manifests as schema `v1` was deprecated when `v2` was released (see [here](https://docs.docker.com/registry/spec/deprecated-schema-v1/)).

In the original submission you reported a problem w.r.t. `v2` manifest, but then followed up on demonstrating the problem on `v1` manifest. Whilst distribution does provide backwards compatibility for v1, for the time being, it's not recommended to continue using `v1` images as the support for `v1` will end up being removed and I believe many other public registries stop supporting `v1` images many moons ago.

**Comment by alexbiehl** [comment]
@milosgajdos Your suggestions sure make sense but I am not in control of the images unfortunately. 

> In the original report you described the issue you were experiencing was when doing PULL; not PUSH

And that is still true. It is the code path for pulling a `v1` manifest that internally makes use of `BlobStore.Put` that violates the invariants. 

> In the original submission you reported a problem w.r.t. v2 manifest, but then followed up on demonstrating the problem on v1 manifest

Interesting, is there a setting that forces `v1`/`v2` usage?
